### ISSUE TITLE
xauth over legacy DISPLAY socket sometimes fails after waking up from system sleep

### ISSUE THREAD

**Issue by telicha** [XQuartz/XQuartz#108]
After I upgraded to 2.8 (no matter what version, beta, rc1, rc2) after coming back from sleep the server doesn't accept any new connections anymore. The old xterm, etc. that  were running before sleep are still working fine, and XQuartz process still seems to be running, but no new connections are allowed, message is 

% xterm
Invalid MIT-MAGIC-COOKIE-1 keyxterm: Xt error: Can't open display: :0

The worst thing is that no matter how I much I try, cannot find all the files touched by XQuartz 2.8, so if I uninstall it, and try to install the old 2.7.11 that was working without any problems whatsoever, it does not run anymore. Thanks for any help either for fixing the issue in future rcs, finding a hack that will let me reconnect to the display, or a way of completely cleaning up so I can go back to 2.7.11. Thanks

**Comment by telicha** [comment]
Is it possible that the issue #102 is related to this issue? The person is not seeing any messages because he is using the GUI instead of command line?

**Comment by jeremyhu** [comment]
To downgrade, uninstall 2.8.0 as per the faq on xquartz.org, then install 2.7.11, then logout and back in.

**Comment by jeremyhu** [comment]
I'm on macOS 11.3 beta, and I'm not seeing this.  I launched XQuartz and opened an xterm with DISPLAY=:0 and another with the launchd socket.

I then put the system to sleep.

I then woke it up about 30s later.

I then used each of those xterms to successfully launch two additional xterms just fine.

**Comment by telicha** [comment]
> To downgrade, uninstall 2.8.0 as per the faq on xquartz.org, then install 2.7.11, then logout and back in.

Thanks for the help ... I executed all the steps on Big Sur 11.2 (20D64) listed in the FAQ: 

% launchctl unload /Library/LaunchAgents/org.xquartz.startx.plist
% sudo launchctl unload /Library/LaunchDaemons/org.xquartz.privileged_startx.plist
% sudo rm -rf /opt/X11* /Library/Launch*/org.xquartz.* /Applications/Utilities/XQuartz.app /etc/*paths.d/*XQuartz
% sudo pkgutil --forget org.xquartz.pkg

(the last one said there is nothing to forget on /).

Then installed again 2.7.11 that was working fine before, logged out and in again, and whenever I try to start XQuartz now complains:


Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY
Application Specific Information:
X.Org X Server 1.18.4 Build Date: 20161025
__glXExtensionInit: AddExtensions failed

So I have to go back and install again 2.8 ...

**Comment by telicha** [comment]
> I'm on macOS 11.3 beta, and I'm not seeing this. I launched XQuartz and opened an xterm with DISPLAY=:0 and another with the launchd socket.
> 
> I then put the system to sleep.
> 
> I then woke it up about 30s later.
> 
> I then used each of those xterms to successfully launch two additional xterms just fine.

Very strange ... I did the same thing as you did and xterms launch just fine after 30 minutes sleep a couple of times. Then left computer sleeping overnight and this morning again can't launch new windows, though the old ones are working fine. Then I tried to uninstall RC2 and re-install 2.7.11 again and had the crash. I guess have to try 2.8 again ...

**Comment by jeremyhu** [comment]
I believe your crash is https://github.com/XQuartz/xorg-server/commit/24f9d825883742fa7ae3086cbea29250af14eb34

That can get triggered if your $DISPLAY isn't correct.  Try rebooting.

**Comment by telicha** [comment]
> I believe your crash is [XQuartz/xorg-server@24f9d82](https://github.com/XQuartz/xorg-server/commit/24f9d825883742fa7ae3086cbea29250af14eb34)
> 
> That can get triggered if your $DISPLAY isn't correct. Try rebooting.

Thanks for the suggestion. Uninstalled again 2.8.0 following FAQ instructions  ... rebooted the machine ...  installed again 2.7.11 ... rebooted the machine again. Still 2.7.11 crashes when I click the Xquartz in Applications/Utilities .... the crash is as in the next paragraph ... 2.7.11 was working fine before I installed 2.8.0 seems that something in 2.8.0 stuck there that causes 2.7.11 to crash at start. I'm pretty sure XQuartz FAQ is missing something. I don't want to re-install Big Sur from scratch ... any suggestions? 2.8 rc2 still has the same problem that randomly after waking up from sleep the Xserver and all the previous windows keep working fine but any new connections are refused with "unable to connect to X server: Connection refused"  no matter what is the setting for $DISPLAY. 

Here is how 2.7.11 crashes: 

Process:               X11.bin [645]
Path:                  /Applications/Utilities/XQuartz.app/Contents/MacOS/X11.bin
Identifier:            org.macosforge.xquartz.X11
Version:               2.7.11 (2.7.112)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           X11.bin [645]
User ID:               502

Date/Time:             2021-03-09 11:46:24.816 -0800
OS Version:            macOS 11.2 (20D64)
Report Version:        12
Anonymous UUID:        29750109-9832-4326-B6DA-2574374CB381


Time Awake Since Boot: 84 seconds

System Integrity Protection: disabled

Crashed Thread:        5

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
X.Org X Server 1.18.4 Build Date: 20161025
__glXExtensionInit: AddExtensions failed
 
abort() called

Thread 0:: Dispatch queue: Client CFPasteboard-AppleInputSourcesInUIOrderPasteboard
0   libsystem_kernel.dylib        	0x00007fff20585e7e mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff205861f0 mach_msg + 60
2   libdispatch.dylib             	0x00007fff20425776 _dispatch_mach_send_and_wait_for_reply + 518
3   libdispatch.dylib             	0x00007fff20425b6e dispatch_mach_send_with_result_and_wait_for_reply + 50
4   libxpc.dylib                  	0x00007fff202fc1ac xpc_connection_send_message_with_reply_sync + 238
....

**Comment by telicha** [comment]
Apparently found a workaround ... so far is working ... even after a short sleep still the Xserver still responds ok. Let's hope tomorrow after sleep keeps working ... 

removed 2.7.11 according to FAQ
rebooted
sudo port install xorg-server
sudo port install xorg-apps
instead of starting clicking on /Applications/Utilities/XQuartz  started using /Applications/MacPorts/X11.app

it seems to use a different set of defaults, so had to go and fix the options but now seems to be working. Says XQuartz 1.20.10 (xorg-server 1.20.10) have no idea what is the equivalence ...

**Comment by jeremyhu** [comment]
The MacPorts build is from the slightly-newer 1.20 branch, and it has a number of crashes that I fixed over the past couple months but haven't patched in MacPorts. Other than that, it's basically exactly what is shipped in XQuartz.

If you're able to provide me with reproduction steps, I'll look into this for you.  As it is, I'm unable to reproduce and thus this ticket is inactionable.  I'll leave it open for a while to see if you can get me data.  If not, I'll end up closing it.

Thanks.

**Comment by jeremyhu** [comment]
I really just need full reproduction steps to reproduce the issue myself and start debugging it.  

1) Does this ever happen with the launchd DISPLAY socket or just :0?

2) How can I reproduce this?  I slept the machine, and everything still seemed fine.

**Comment by telicha** [comment]
I'm so sorry I have no idea what is a launchd DISPLAY socket... I can retrace the steps one by one:

a) Running Big Sur 11.2 (20D64) on iMac19,1
b) Installed Big Sur over a clean Catalina install that included installing Xquartz 2.7.11 and setting a few X11 preferences
c) Everything worked perfect until end of February, including sleeping, waking, adding new windows after sleep, etc. 
d) Usually run a few xterm windows and xemacs occasionally CrossVC to check in code in CVS, always started X11 after a reboot by running /Applications/Utilities/XQuartz (maybe that is related with the launchd DISPLAY setting?) but withing X11 running 

% echo $DISPLAY 
 :0 

e) End of February machine was running for about a month about 40 cycles of sleep/wake about 15h/day wake, so decided was a good idea to reboot, and when shutting down cleanly X11 server before the reboot, asked me if I wanted to upgrade to 2.8.0 I think was rc1 or rc2, and I accepted the upgrade without thinking much about it, installed, finished rebooting the machine. 
f) Used a couple of days normally - there was at least one sleep that the server has a problem after wake - the strange thing is that the existing windows keep working normally. But if I got to one of the xterm and do:

% xterm
Invalid MIT-MAGIC-COOKIE-1 keyxterm: Xt error: Can't open display: :0

The strange thing is that Xquartz process is still running normally, and  /tmp/.X0-lock is there and has the right process number, and the socket /tmp/.X11-unix/X0 seems to be fine. It is just that when any application tries to call XOpenDisplay() it seems to fail right there. The display keeps running without any problems, it is just that NEW X applications cannot initiate anymore - all the stuff that existed before it went to sleep remain working without any problems. 

g) Tried to roll back to 2.7.11 after all it was working fine, thought there was a problem with the release candidates. But no matter how I get rid of 2.8.0 - this morning even rebooted after following the FAQ instructions - when I install 2.7.11 log out, reboot etc always go back to the problem of  Xquartz crashing right away and if I try to generate a report to send to Apple always get the same 

__glXExtensionInit: AddExtensions failed
abort() called

The truth is that maybe when installing 2.7.11 on Big Sur it never works no matter how hard I try to scrape everything from 2.8.0 and reboot/log out, etc...   I also noticed that during the installation of 2.7.11 it asked me something about the installer wants to have control into events and I always reply "yes" ... that question was never there when I installed originally the working 2.7.11 but that was several months ago when I installed Catalina from scratch. On Catalina Xquartz worked without any issues whatsoever, and upgraded to Big Sur and then started XQuartz and all worked normally as well. 

Apparently the same problem persists on the Macports server ... I had one cycle of sleep/wake and everything worked normal, but then when the display itself went to sleep and I had to login again, then again have the same issue, all xterm, xemacs, etc. that were there before remain connected without any problems, but whenever I try to do:

% xterm
Invalid MIT-MAGIC-COOKIE-1 keyxterm: Xt error: Can't open display: :0

Quite bizarre behavior .... I think the real question is why after presenting login credentials a running X11 server keeps working normal but no longer accepts new connections? And why it says Invalid MIT-MAGIC-COOKIE-1 keyxterm? Could be that the problem is not related with X but some authentication issue? How can I check that? The command line for the actual server looks like this (ps auxww)

USER               PID  %CPU %MEM      VSZ    RSS   TT  STAT STARTED      TIME COMMAND
ricardo           4036   0.0  0.0  4410964   1684   ??  S    12:37PM   0:00.02 /opt/local/bin/Xquartz :0 -nolisten tcp -iglx -auth /Volumes/share/ricardo/.serverauth.3860

Could be that now my .serverauth file no longer matches something that was renewed when I typed my password to unlock the display? Can I test it without checking the server authorization maybe the problem will go away?

Thanks!!!

**Comment by jschnack-uni** [comment]
Dear jeremyhu,

the mentioned error appeared first after last weeks big sur update:
Running without a11y support!
Invalid MIT-MAGIC-COOKIE-1 keyUnable to init server: Broadway-Bildtyp nicht unterstützt: »localhost:11.0«
Error: cannot open display: localhost:11.0

If I want to start some x-application from any of my linux servers.

Last week switching off and on XQuartz helped although I never really understood why.

After today's update of big rur (11.2.3) I don't get it running. 

It worked for ages: ssh - Y ...
I was always very satisfied. I don't know what they changed in big sur.

Thank you for all of your work, Juergen

**Comment by jschnack-uni** [comment]
Found on some other web page:
ssh -Y ...  -o "XAuthLocation=/opt/X11/bin/xauth"
after closing and restarting XQuartz worked.

Then it works, I only get some warmings:
Running without a11y support!
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
But this has been the case all the time.

Is this no longer set on big sur, was it set before?

**Comment by telicha** [comment]
Thanks all for your comments and help. Sorry for misinterpreting some errors because they have been buried by "sticky states" and confused me. Here is a summary of the issues found:

a) If you try to install Xquartz 2.7.11 in Big Sur (>=11.2?) something is missing during the installation that always crashes Xquartz at startup:

__glXExtensionInit: AddExtensions failed

it is important to notice that if you installed it previously (I did after Catalina install) it will continue to work fine, which makes this issue a bit harder to debug. 

b) Something changed either in Big Sur or in Xquartz since 2.7.11 (and also on Macports) that when "Authenticate Connections"  is set (the default in Preferences->Security)  and a person has to present new credentials - either the Mac sleeps or just the display sleeps and one needs to unlock it - the original credentials created when the server starts in ~/.serverauth.xxxxx (I think but not sure specifically the MIT-MAGIC-COOKIE) becomes invalid. This is a serious bug that manifests in many different and confusing ways as shown by other users.  No matter what happens and how many times the Xserver needs to be reset, it should always accept as a valid authorization the file provided in -auth command line option (as long as the file does not change itself I guess). 

In a nutshell, that is what is going on. I'm not sure what actions are necessary to fix (a) or (b) - perhaps a new ticket is needed, but I consider this case closed, just pending Jeremy's comment on how to proceed forward. Thanks you all!

**Comment by jeremyhu** [comment]
> __glXExtensionInit: AddExtensions failed

This usually indicates that you're running 2.7.x with the 2.8.0 LaunchAgent still loaded.  Make sure you remove it (see https://www.xquartz.org/FAQs.html) and reboot.  This particular crash was fixed in one of the 2.8.0 betas, but that doesn't help you on 2.7.11.

> % echo $DISPLAY
> :0

Do you know why that is?  It should look like:

$ echo $DISPLAY
/private/tmp/com.apple.launchd.drREC61dJT/org.xquartz:0

That's the launchd socket.  Are you explicitly setting :0?  Check the FAQ for debugging tips around incorrect DISPLAY envvars.

> ssh -Y ... -o "XAuthLocation=/opt/X11/bin/xauth"

The XQuartz post-installation script should be setting that up for you in /etc/sshd_config (or /etc/ssh/sshd_config).  Did it not?

**Comment by jeremyhu** [comment]
So I think the main issue here is figuring out why the auth token somehow becomes invalid across those sleep/wake cycles.

The ideal scenario is that you could provide me with some steps to reproduce this (ie: we could figure out exactly what causes the invalidation), so I could reproduce myself and debug.

**Comment by telicha** [comment]
> So I think the main issue here is figuring out why the auth token somehow becomes invalid across those sleep/wake cycles.
> 
> The ideal scenario is that you could provide me with some steps to reproduce this (ie: we could figure out exactly what causes the invalidation), so I could reproduce myself and debug.

Thanks Jeremy for your help. Here are the steps - on Big Sur (>=11.2) and using 2.8.0 XQuartz:

a) Make sure you set on "Authenticate Connections"  on  XQuartz->Preferences->Security. If it was not on, turn it on and restart the X server. 

b) Not sure if it makes any difference or not, but I found out that "sneaking behind me" the macports distribution added a setenv DISPLAY :0  in my .cshrc file

c) If it was successful, when you do ps auxww and look for XQuartz the entry will look something like this - notice the command line for the Xquartz includes the -auth: 

USER PID %CPU %MEM VSZ RSS TT STAT STARTED TIME COMMAND
ricardo 4036 0.0 0.0 4410964 1684 ?? S 12:37PM 0:00.02 /opt/local/bin/Xquartz :0 -nolisten tcp -iglx -auth /Volumes/share/ricardo/.serverauth.3860

d) You can check also for the file ˜/.serverauth.xxxxx - usually it has one cookie ... make keep a copy of the contents of that file

e) start an xterm 

f) put the computer to sleep, or the display to sleep a few times, make sure you need to type in a password to unblock it ... in case your settings don't require it, make sure Security&Privacy -> General has "Require password" set

g) inside the xterm try to start another xterm, most likely you'll get something like:

Invalid MIT-MAGIC-COOKIE-1 keyxterm: Xt error: Can't open display: :0

**Comment by jeremyhu** [comment]
> macports distribution added a setenv DISPLAY :0 in my .cshrc file

Yeah, that was a bug in MacPorts that has since been fixed.  It saw that ".0" in "11.0" and thought you were on OS X 10.0... =/

The repro steps above are basically what I tried before, but I'll try them again. Admitedly, I only tried about three times.

**Comment by jeremyhu** [comment]
Yeah, I'm not seeing this.  I'm on an Apple Silicon MacBook Air running macOS 11.3 Beta.  I sleep and come back and DISPLAY=:0 still continues to work just fine...

**Comment by jeremyhu** [comment]
That error is thrown in:

XID
MitCheckCookie(unsigned short data_length,
               const char *data, ClientPtr client, const char **reason)
{
    struct auth *auth;

    for (auth = mit_auth; auth; auth = auth->next) {
        if (data_length == auth->len &&
            timingsafe_memcmp(data, auth->data, (int) data_length) == 0)
            return auth->id;
    }
    *reason = "Invalid MIT-MAGIC-COOKIE-1 key";
    return (XID) -1;
}

---

I wonder if you're seeing the server regenerate for some reason, and the next generation isn't setting up auth.

Do you see anything of note in ~/Library/Logs/X11/org.xquartz.log when this is happening?

**Comment by telicha** [comment]
Hello Jeremy,

Checked the log files (did not know they existed)  and the only error message that appears on all of them but doesn't seem to matter is about error loading keymap /tmp/server-0.xkm 

I also noted that when the server is started with command line arguments -auth ˜/.serverauth.xxxx that file only seems to contain one MIT cookie but without these command line arguments there is something called ˜/.Xauthority that contains 7 magic cookies, seems they are related to the different internet channels I have lo0, en0, en1, etc.  because one of them has my fully-qualified hostname ... maybe that will help ... it is possible sometimes the data ptr in the code above  gets generated from a different internet channel - for instance if wifi is still off - and doesn't match the specific single cookie in ˜/.serverauth but does match one of the cookies in .Xauthority? 
An evidence of that would be this text: 


![Screen Shot 2021-03-11 at 10 29 25 AM](https://user-images.githubusercontent.com/27403312/110836241-b0341880-8254-11eb-8866-5c3eeb5e7ca8.png)

The other possibility is that there is some kind of fluke with IP addresses in 11.2 and was fixed in 11.3 that's why you don't see the problem. 

Since my network is pretty safe, and I'm a single user in this computer and don't allow remote connections, I don't really need to set "Authenticate connections"  - it doesn't allow any connections from other hosts anyway - and if the problem goes away in 11.3 I guess you could close the ticket, I think we have a pretty understanding of what's going on and how to fix it for other people that see the same problem until 11.3 becomes official, right?

**Comment by telicha** [comment]
Maybe this is "bad karma" because in the late 80s I abused the "Authenticate connections" by logging into my buddies computers (we didn't have computers in our desks, we shared the different computers in a large room) in school and running something called xmelt that lasted about 5 seconds and looked like the screen was melting ... it was harmless banter, young kids pranks that were harmless, but would be really annoying specially around 3am he he he :-)

**Comment by jeremyhu** [comment]
~/.serverauth.* contains the auth keys for that specific server.
~/.Xauthority contains the auth keys for all known servers.

:0 should not be using the network.  It should be using the unix domain socket, so your entries with your network addresses shouldn't be relevant there.

**Comment by jeremyhu** [comment]
Closing due to lack of information.  Please re-open if you can help triage this or have a way for me to reproduce.
